docs(v2): Create upgrade guide and versioning policy (#1856)
aws-powertools/powertools-lambda-java: docs/upgrade.md

We've made minimal breaking changes to make your transition to `v2` as smooth as possible.

### Quick summary

The following table shows a summary of the changes made in `v2` and whether code changes are necessary. Each change that requires a code change links to a section below explaining more details.

| Area                 | Change                                                                                                                                                                                             | Code change required |
| -------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------- |
| **Logging**          | The [logging module was re-designed](#redesigned-logging-utility) from scratch to support popular Java logging paradigms and libraries like `log4j2`, `logback`, and `slf4j`.                      | Yes                  |

> | **Metrics**          | [Changed public interface](#updated-metrics-utility-interface) to remove direct coupling with `aws-embedded-metrics-java`. | Yes                  |

Reviewer: ```suggestion
| **Metrics**          | [Changed public interface](#updated-metrics-utility-interface) to remove direct coupling with `aws-embedded-metrics-java`.                                                                          | Yes                  |
```

> | **Metrics**          | [Changed public interface](#updated-metrics-utility-interface) to remove direct coupling with `aws-embedded-metrics-java`.                                                                          | Yes                  |

Reviewer: ```suggestion
| **Metrics**          | [Changed public interface](#updated-metrics-utility-interface) to remove direct coupling with `aws-embedded-metrics-java`.                                                                         | Yes                  |
```

    private final BatchMessageHandler<SQSEvent, SQSBatchResponse> handler;

    public SqsBatchHandler() {
        handler = new BatchMessageHandlerBuilder()
                .withSqsBatchHandler()
                .buildWithMessageHandler(this::processMessage, Product.class);
    }

    @Override
    @Validation(inboundSchema = "classpath:/schema_in.json", outboundSchema = "classpath:/schema_out.json")
    public SQSBatchResponse handleRequest(SQSEvent sqsEvent, Context context) {
        return handler.processBatch(sqsEvent, context);
    }

    private void processMessage(Product p, Context c) {
        // Process the product
    }
}
```

- **Prior to `v2`** this caused the whole batch to fail.
- **After `v2`** this will add only the failed events to the batch item failure list in the response and process the remaining messages.

!!! info "Check if your workload can tolerate this behavior and make sure it is designed for idempotency when using partial batch item failures. We offer the [Idempotency](./utilities/idempotency.md) utility to simplify integration of idempotent behavior in your workloads."

## AspectJ runtime not included by default anymore

> The AspectJ runtime is no longer included as a transitive dependency of Powertools. For all utilities offering annotations using AspectJ compile-time weaving, you need to include the AspectJ runtime yourself now. This is also called out with a complete example in our [installation guide](./index.md). For Maven projects, make sure to add the following dependency in your dependencies section:

Reviewer 2: ```suggestion
The AspectJ runtime is no longer included as a transitive dependency of Powertools. For all utilities offering annotations using AspectJ compile-time weaving, you need to include the AspectJ runtime yourself now. This is also documented, with a complete example, in our [installation guide](./index.md). For Maven projects, make sure to add the following dependency in your dependencies section:
```

```diff
+ <dependency>
+     <groupId>org.aspectj</groupId>
+     <artifactId>aspectjrt</artifactId>
+     <version>1.9.22</version>
+ </dependency>
```

## Removed `powertools-sqs` module in favor of `powertools-batch`

The archived documentation contains a migration guide for both large message handling using `powertools-sqs` and batch processing using `powertools-sqs`. The sections below explain the high-level steps for your convenience.

### Migrating SQS Batch processing (`@SqsBatch`)

> The [batch processing library](./utilities/batch.md) provides a way to process messages and gracefully handle partial failures for SQS, Kinesis Streams, and DynamoDB Streams batch sources. In comparison to the legacy SQS Batch library, it relies on [Lambda partial batch responses](https://docs.aws.amazon.com/lambda/latest/dg/services-sqs-errorhandling.html#services-sqs-batchfailurereporting){target="\_blank"}, which allows the library to provide a simpler, reliable interface for processing batches.

Reviewer 2: ```suggestion
The [batch processing library](./utilities/batch.md) provides a way to process messages and gracefully handle partial failures for SQS, Kinesis Streams, and DynamoDB Streams batch sources. In comparison to the legacy SQS Batch library, it relies on [Lambda partial batch responses](https://docs.aws.amazon.com/lambda/latest/dg/services-sqs-errorhandling.html#services-sqs-batchfailurereporting){target="\_blank"}, which allows the library to provide a simpler, more reliable interface for processing batches.
```

<!-- prettier-ignore-start -->
!!! info "Some tuneables from `powertools-sqs` are no longer provided."
    - **Non-retryable Exceptions** - there is no mechanism to indicate in a partial batch response that a particular message
        should not be retried and instead moved to DLQ - a message either succeeds, or fails and is retried. A message
        will be moved to the DLQ once the normal retry process has expired.
    - **Suppress Exception** - The new batch processor does not throw an exception on failure of a handler. Instead,
        its result must be returned by your code from your message handler to Lambda, so that Lambda can manage
        the completed messages and retry behaviour.
<!-- prettier-ignore-end -->

### Migrating SQS Large message handling (`@SqsLargeMessage`)

- Replace the dependency in Maven / Gradle: `powertools-sqs` ==> `powertools-large-messages`
- Replace the annotation: `@SqsLargeMessage` ==> `@LargeMessage` (the new module handles both SQS and SNS)
- Move the annotation away from the Lambda `handleRequest` method and put it on a method with `SQSEvent.SQSMessage` or `SNSEvent.SNSRecord` as first parameter.

> - The annotation now handles a single message, contrary to the previous version that was handling the complete batch. It gives more control, especially when dealing with partial failures with SQS (see the batch module).

Reviewer 2: ```suggestion
- The annotation now handles a single message, contrary to the previous version that was handling the complete batch. This gives more control, especially when dealing with partial failures with SQS (see the batch module).
```

> - The new module only provides an annotation, an equivalent to the `SqsUtils` class is not available anymore in this new version.

Reviewer 2: ```suggestion
- The new module only provides an annotation: an equivalent to the `SqsUtils` class is not available anymore in this new version.
```
